Implements FETCH_GROUP

Indigo744 opened the pull request:
Fix #817 

Implements FETCH_GROUP and works with FETCH_ASSOC, FETCH_NUM and FETCH_BOTH.

Note that it only make sense to implement in `fetchAll`, as `fetch` only returns one row (but it is checked in the test case to be sure).

jakubmisek commented:
thanks for this :) I wasn't sure how to approach this weird PDO flag.

Although the code seems a bit unclean :)) and internally each of those array operations is very expensive also. Let me think about it.

jakubmisek commented:
I think FETCH_GROUP is the flag (determined above `var flags = style & PDO_FETCH.Flags`) that should be handled separately after the `result` is populated. So the current switch code block stays intact.

```c#
if (flags == PDO_FETCH.FETCH_GROUP) {
 // transform result array into a new result array
}
```

Indigo744 commented:
I moved the code outside the switch, as requested. Cleaner indeed.

I changed the way numeric keys are determined. I actually check the first row to see if there is a numeric key. I think it's better than checking all flags, since it'll work even if a new flag is added.

In any case, using FETCH_GROUP will always be more expansive than without, but will still be faster than doing it in PHP!

jakubmisek commented:
good! May I ask for write access to your `FETCH_GROUP` branch? I'd add some cleanups

Indigo744 commented:
> May I ask for write access to your `FETCH_GROUP` branch? I'd add some cleanups

I'm not sure how? I can see _Allow edits by maintainers_ and it is already checked so maybe that's it?

jakubmisek commented:
> I'm not sure how? I can see _Allow edits by maintainers_ and it is already checked so maybe that's it?

I'm probably not listed within your project *maintainers* group

Indigo744 commented:
Well, that shouldn't be needed, as the docs says about _Allow edits by maintainers_:
> If checked, users with write access to peachpiecompiler/peachpie can add new commits to your FETCH_GROUP branch. You can always change this setting later.

So that should be good.

jakubmisek commented:
@Indigo744 thanks :)

Indigo744 commented on src/PDO/Peachpie.Library.PDO/PDOStatement.cs:
```diff
@@ -485,40 +485,63 @@ public virtual PhpArray fetchAll(PDO_FETCH fetch_style = default, PhpValue fetch
-                    }
+            // hasNumKeys: when FETCH_NUM or FETCH_BOTH
+            var hasNumKeys =
+                (style & ~PDO_FETCH.Flags) == PDO_FETCH.FETCH_NUM ||
```
Beware for hasNumKeys, you need to also check for `FETCH_DEFAULT` since FETCH_BOTH is the default...

It was partly why I choose to go to the "check if there is a numeric key in the first row" route...

jakubmisek replied:
good point!
